google/cloud/storage/bucket_metadata.cc

```diff
@@ -384,6 +384,271 @@ std::ostream& operator<<(std::ostream& os, BucketMetadata const& rhs) {
   return os << "}";
 }
 
+std::string BucketMetadataPatchBuilder::BuildPatch() const {
+  internal::PatchBuilder tmp = impl_;
+  if (labels_subpatch_dirty_) {
+    if (labels_subpatch_.empty()) {
+      tmp.RemoveField("labels");
+    } else {
+      tmp.AddSubPatch("labels", labels_subpatch_);
+    }
+  }
+  return tmp.ToString();
+}
+
+BucketMetadataPatchBuilder& BucketMetadataPatchBuilder::SetAcl(
+    std::vector<BucketAccessControl> const& v) {
+  if (v.empty()) {
+    return ResetAcl();
+  }
+  std::vector<internal::nl::json> array;
+  array.reserve(v.size());
+  for (auto const& a : v) {
+    array.emplace_back(internal::nl::json{
+        {"entity", a.entity()},
+        {"role", a.role()},
+    });
+  }
+  impl_.SetArrayField("acl", array);
+  return *this;
+}
+
+BucketMetadataPatchBuilder& BucketMetadataPatchBuilder::ResetAcl() {
+  impl_.RemoveField("acl");
```

Comment:
Also just remembered that ResetAcl actually won't work :(  See https://issuetracker.google.com/issues/67861222#comment2.

Even if we had a way to tell the API, "Yes, I really meant *null* or *empty JSON list*, and you should clear all the entries from my ACL", there's still the caveat that GCS will always add back the original owner entity to the bucket (or object, if operating on an object), as mentioned at [1]:

"When you apply a new ACL to a bucket or object, Cloud Storage respectively adds OWNER permission to the bucket or object owner if you omit the grants. It does not grant the project owners group OWNER permission for an object (unless the object was created by an anonymous user), so you must explicitly include it."

The only workaround I'm aware of is:
1) List ObjectAccessControls for the bucket [2]
2) Loop over the entities, calling delete [3] on each one.
3) For step 2, ignore the case where we get a 403 back with the error message "The owner of the resource is required to have OWNER access." (from trying to remove the owner entity).

The same workaround exists for defaultObjectAcls, but we don't have to worry about step 3 since the owner entity can be removed from that list.

Personally, I find the above workaround to be pretty ugly and have several downsides (mainly that it takes multiple requests for one "operation").  I'd recommend not even adding in the `ResetAcl` or `ResetDefaultObjectAcl` methods until https://issuetracker.google.com/issues/67861222#comment16 (or whatever proposed fix is chosen) is implemented.

[1] https://cloud.google.com/storage/docs/access-control/lists
[2] https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/list
[3] https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/delete

Reply:
Hmm... the client library (I think) should "send the request that the user wants", if the server has a bug that is not something we want to workaround in the library.  We should probably put a warning in the documentation though.

BTW, do you have a externally visible link to these issues?

Reply:
Sounds fair.

The add-the-owner-entity-back-in issue is documented at [1].  As for the clear-doesnt-actually-clear-for-acls-in-patch issue, no, there isn't a link for a publicly visible issue that I know of.

Reply:
I created #998 to fix the documentation, it will need to be in more places than just this class.